## Import libraries and dataframe

In [1]:
# import libraries
import numpy as np
import pandas as pd
import os

In [2]:
# paert 1: import data frame 
path = r'C:\Users\98912\MAY 2022 Instacart Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02-Data', 'Prepared Data', 'orders_products_merged_ver2.pkl'))

In [3]:
ords_prods_merge.shape

(32404859, 18)

In [4]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

In [5]:
# part 2: aggregated mean of the “order_number” column grouped by “department_id”
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Part 3: according to this results the department id 21 which is equivalent to 'missing' has the highest frequency while department id 5 which is 'alcohol' has the lowest mean of order number. Also the subset didn't include the whole department ids. So we can say that the sub-sample was not representative of all the population.


### Addititional note: I think that the insight developed about this part of the analysis in this exercise is not logical. Because the 'order_number' is not the number of items bought by the user, but it is the sequence number of each order for each user_id as the first oreder, second order and etc. . As we check it, we can see that the order_number starts with 1 for all users and have a sequencial format and moreover for orders with 'order_number' of for example '1' we have more than 1 item in the shoping order. As a result I think this insight that mean of 'order_number' for each department is a sign of success or fail is not relevent. 

In [11]:
ords_prods_merge[(ords_prods_merge['user_id']==1) & (ords_prods_merge['order_number']==1)]

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0       2539329        1             1                  2                  8   
35791   2539329        1             1                  2                  8   
51726   2539329        1             1                  2                  8   
58202   2539329        1             1                  2                  8   
60725   2539329        1             1                  2                  8   

       days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                         0.0         196                  1          0   
35791                     0.0       14084                  2          0   
51726                     0.0       12427                  3          0   
58202                     0.0       26088                  4          0   
60725                     0.0       26405                  5          0   

                                  product_name  aisle_id  department_id  \
0                                         Soda        77              7   
35791  Organic Unsweetened Vanilla Almond Milk        91             16   
51726                      Original Beef Jerky        23             19   
58202               Aged White Cheddar Popcorn        23             19   
60725         XL Pick-A-Size Paper Towel Rolls        54             17   

       prices _merge    price_range_loc     busiest_day    busiest_days  \
0         9.0   both  Mid-range product  Regularly busy  Regularly busy   
35791    12.5   both  Mid-range product  Regularly busy  Regularly busy   
51726     4.4   both  Low-range product  Regularly busy  Regularly busy   
58202     4.7   both  Low-range product  Regularly busy  Regularly busy   
60725     1.0   both  Low-range product  Regularly busy  Regularly busy   

      busiest_period_of_day  
0            Average orders  
35791        Average orders  
51726        Average orders  
58202        Average orders  
60725        Average orders

### Part 4: Create loyalty flag

In [12]:
# first define the new column as 'max_order' to show the maximum order_number using groupby and transform 
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform(np.max)

In [13]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Regularly busy        Average orders         10  
1  Regularly busy    Slowest days        Average orders         10  
2  Regularly busy    Slowest days           Most orders         10  
3      Least busy    Slowest days        Average orders         10  
4      Least busy    Slowest days           Most orders         10

In [14]:
# create loyalty_flag column using .loc
ords_prods_merge.loc[ords_prods_merge['max_order']>40, 'loyalty_flag'] = 'Loyal_customer'
ords_prods_merge.loc[(ords_prods_merge['max_order']>10) & (ords_prods_merge['max_order']<=40), 'loyalty_flag'] = 'Regular_customer'
ords_prods_merge.loc[ords_prods_merge['max_order']<=10, 'loyalty_flag'] = 'New_customer'

In [15]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days           Most orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  
0  New_customer  
1  New_customer  
2  New_customer  
3  New_customer  
4  New_customer

In [16]:
ords_prods_merge['loyalty_flag'].value_counts()

Regular_customer    15876776
Loyal_customer      10284093
New_customer         6243990
Name: loyalty_flag, dtype: int64

In [17]:
# part 5: compare the shoping habits of different customers

ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'max', 'min']})

prices              
                       mean      max  min
loyalty_flag                             
Loyal_customer    10.386336  99999.0  1.0
New_customer      13.294670  99999.0  1.0
Regular_customer  12.495717  99999.0  1.0

### The maximum price of this table cannot be used for comparison because the number '99999' is not a real number and it has been used probably for 'missing values'. However, when we compare the average prices we can say that in average loyal customers by cheaper items comparing to new customers. 

### Part 6: create spending flag for each user based on mean price of ordered items

In [18]:
# first create new column 'mean_order-price' by using groupby and transform
ords_prods_merge['mean_prder_price'] = ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy    Slowest days        Average orders        10   
2  Regularly busy    Slowest days           Most orders        10   
3      Least busy    Slowest days        Average orders        10   
4      Least busy    Slowest days           Most orders        10   

   loyalty_flag mean_prder_price  
0  New_customer         6.367797  
1  New_customer         6.367797  
2  New_customer         6.367797  
3  New_customer         6.367797  
4  New_customer         6.367797  

[5 rows x 21 columns]

In [19]:
# second define the 'spending_flag' using .loc
ords_prods_merge.loc[ords_prods_merge['mean_prder_price']<10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['mean_prder_price']>=10, 'spending_flag'] = 'High spender'
ords_prods_merge.head(20)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      0.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  prices  _merge    price_range_loc     busiest_day  \
0          Soda  ...     9.0    both  Mid-range product  Regularly busy   
1          Soda  ...     9.0    both  Mid-range product  Regularly busy   
2          Soda  ...     9.0    both  Mid-range product  Regularly busy   
3          Soda  ...     9.0    both  Mid-range product      Least busy   
4          Soda  ...     9.0    both  Mid-range product      Least busy   
5          Soda  ...     9.0    both  Mid-range product  Regularly busy   
6          Soda  ...     9.0    both  Mid-range product  Regularly busy   
7          Soda  ...     9.0    both  Mid-range product  Regularly busy   
8          Soda  ...     9.0    both  Mid-range product  Regularly busy   
9          Soda  ...     9.0    both  Mid-range product      Least busy   
10         Soda  ...     9.0    both  Mid-range 

In [35]:
# number of each customer type
from statistics import mode
ords_prods_merge.groupby('user_id')['spending_flag'].agg(lambda x: mode(x)).value_counts()

Low spender     200852
High spender      5357
Name: spending_flag, dtype: int64

### part 7: create 'frequency flag

In [36]:
# first determine the median of 'days_since_prior_order' for each user_id
ords_prods_merge['order_frequency'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform(np.median)
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge    price_range_loc     busiest_day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy        Average orders        10  New_customer   
1    Slowest days        Average orders        10  New_customer   
2    Slowest days           Most orders        10  New_customer   
3    Slowest days        Average orders        10  New_customer   
4    Slowest days           Most orders        10  New_customer   

  mean_prder_price  spending_flag order_frequency  
0         6.367797    Low spender            20.0  
1         6.367797    Low spender            20.0  
2         6.367797    Low spender            20.0  
3         6.367797    Low spender            20.0  
4         6.367797    Low spender            20.0  

[5 rows x 23 columns]

In [38]:
# define the flag using .loc
ords_prods_merge.loc[ords_prods_merge['order_frequency']> 20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['order_frequency']<=20) & (ords_prods_merge['order_frequency']>10) , 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['order_frequency']<=10, 'frequency_flag'] = 'Frequent customer'

ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day    busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy    Slowest days   
2         Soda  ...  Mid-range product  Regularly busy    Slowest days   
3         Soda  ...  Mid-range product      Least busy    Slowest days   
4         Soda  ...  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag mean_prder_price  \
0        Average orders        10  New_customer         6.367797   
1        Average orders        10  New_customer         6.367797   
2           Most orders        10  New_customer         6.367797   
3        Average orders        10  New_customer         6.367797   
4           Most orders        10  New_customer         6.367797   

  spending_flag  order_frequency    frequency_flag  
0   Low spender             20.0  Regular customer  
1   Low spender             20.0  Regular customer  
2   Low spender             20.0  Regular customer  
3   Low spender             20.0  Regular customer  
4   Low spender             20.0  Regular customer  

[5 rows x 24 columns]

In [39]:
# number of each customer type
ords_prods_merge.groupby('user_id')['frequency_flag'].agg(lambda x: mode(x)).value_counts()

Frequent customer        103446
Regular customer          59484
Non-frequent customer     43279
Name: frequency_flag, dtype: int64

In [41]:
# Export newly modified datafram
ords_prods_merge.to_pickle(os.path.join(path, '02-Data', 'Prepared Data', 'orders_products_merged_ver3.pkl'))